File: BikeHaus.Client/src/app/services/statistics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface DailyStats {
  date: string;
  purchaseCount: number;
  saleCount: number;
  expenseCount: number;
  purchaseAmount: number;
  saleAmount: number;
  expenseAmount: number;
  dailyProfit: number;
  dailyNetProfit: number;
}

export interface TopBrand {
  brand: string;
  count: number;
  totalRevenue: number;
}

export interface ExpenseByCategory {
  category: string;
  count: number;
  totalAmount: number;
}

export interface Statistics {
  startDate: string;
  endDate: string;
  purchaseCount: number;
  saleCount: number;
  expenseCount: number;
  totalPurchaseAmount: number;
  totalSaleAmount: number;
  totalExpenseAmount: number;
  profit: number;
  netProfit: number;
  averagePurchasePrice: number;
  averageSalePrice: number;
  averageProfit: number;
  dailyBreakdown: DailyStats[];
  topBrands: TopBrand[];
  expensesByCategory: ExpenseByCategory[];
}

@Injectable({
  providedIn: 'root',
})
export class StatisticsService {
  private apiUrl = `${environment.apiUrl}/statistics`;

  constructor(private http: HttpClient) {}

  getStatistics(startDate: Date, endDate: Date): Observable<Statistics> {
    const params = new HttpParams()
      .set('startDate', startDate.toISOString())
      .set('endDate', endDate.toISOString());
    return this.http.get<Statistics>(this.apiUrl, { params });
  }

  getTodayStatistics(): Observable<Statistics> {
    return this.http.get<Statistics>(`${this.apiUrl}/today`);
  }

  getWeekStatistics(): Observable<Statistics> {
    return this.http.get<Statistics>(`${this.apiUrl}/week`);
  }

  getMonthStatistics(): Observable<Statistics> {
    return this.http.get<Statistics>(`${this.apiUrl}/month`);
  }

  getQuarterStatistics(): Observable<Statistics> {
    return this.http.get<Statistics>(`${this.apiUrl}/quarter`);
  }

  getYearStatistics(): Observable<Statistics> {
    return this.http.get<Statistics>(`${this.apiUrl}/year`);
  }
}

File: BikeHaus.Application/Services/StatisticsService.cs
using BikeHaus.Application.DTOs;
using BikeHaus.Application.Interfaces;
using BikeHaus.Domain.Interfaces;

namespace BikeHaus.Application.Services;

public class StatisticsService : IStatisticsService
{
    private readonly IPurchaseRepository _purchaseRepository;
    private readonly ISaleRepository _saleRepository;
    private readonly IExpenseRepository _expenseRepository;

    public StatisticsService(
        IPurchaseRepository purchaseRepository,
        ISaleRepository saleRepository,
        IExpenseRepository expenseRepository)
    {
        _purchaseRepository = purchaseRepository;
        _saleRepository = saleRepository;
        _expenseRepository = expenseRepository;
    }

    public async Task<StatisticsDto> GetStatisticsAsync(DateTime startDate, DateTime endDate)
    {
        // Ensure end date includes the whole day
        var endDateInclusive = endDate.Date.AddDays(1).AddTicks(-1);

        var allPurchases = await _purchaseRepository.GetAllAsync();
        var allSales = await _saleRepository.GetAllAsync();
        var allExpenses = await _expenseRepository.GetAllAsync();

        // Filter by date range
        var purchases = allPurchases
            .Where(p => p.Kaufdatum >= startDate.Date && p.Kaufdatum <= endDateInclusive)
            .ToList();

        var sales = allSales
            .Where(s => s.Verkaufsdatum >= startDate.Date && s.Verkaufsdatum <= endDateInclusive)
            .ToList();

        var expenses = allExpenses
            .Where(e => e.Datum >= startDate.Date && e.Datum <= endDateInclusive)
            .ToList();

        var totalPurchaseAmount = purchases.Sum(p => p.Preis);
        var totalSaleAmount = sales.Sum(s => s.Preis);
        var totalOperationalExpenseAmount = expenses.Sum(e => e.Betrag);
        var totalExpenseAmount = totalOperationalExpenseAmount + totalPurchaseAmount;
        var profit = totalSaleAmount - totalPurchaseAmount;
        var netProfit = totalSaleAmount - totalExpenseAmount;

        // Calculate averages
        var avgPurchase = purchases.Count > 0 ? totalPurchaseAmount / purchases.Count : 0;
        var avgSale = sales.Count > 0 ? totalSaleAmount / sales.Count : 0;
        var avgProfit = sales.Count > 0 ? profit / sales.Count : 0;

        // Daily breakdown
        var days = (endDate.Date - startDate.Date).Days + 1;
        var dailyBreakdown = Enumerable.Range(0, days)
            .Select(d =>
            {
                var date = startDate.Date.AddDays(d);
                var dayPurchases = purchases.Where(p => p.Kaufdatum.Date == date).ToList();
                var daySales = sales.Where(s => s.Verkaufsdatum.Date == date).ToList();
                var dayExpenses = expenses.Where(e => e.Datum.Date == date).ToList();
                var dayPurchaseAmount = dayPurchases.Sum(p => p.Preis);
                var daySaleAmount = daySales.Sum(s => s.Preis);
                var dayOperationalExpenseAmount = dayExpenses.Sum(e => e.Betrag);
                var dayExpenseAmount = dayOperationalExpenseAmount + dayPurchaseAmount;
                var dayProfit = daySaleAmount - dayPurchaseAmount;

                return new DailyStatsDto(
                    Date: date,
                    PurchaseCount: dayPurchases.Count,
                    SaleCount: daySales.Count,
                    ExpenseCount: dayExpenses.Count + dayPurchases.Count,
                    PurchaseAmount: dayPurchaseAmount,
                    SaleAmount: daySaleAmount,
                    ExpenseAmount: dayExpenseAmount,
                    DailyProfit: dayProfit,
                    DailyNetProfit: daySaleAmount - dayExpenseAmount
                );
            })
            .ToList();

        // Top brands by sales
        var topBrands = sales
            .Where(s => s.Bicycle != null)
            .GroupBy(s => s.Bicycle!.Marke)
            .Select(g => new TopBrandDto(
                Brand: g.Key,
                Count: g.Count(),
                TotalRevenue: g.Sum(s => s.Preis)
            ))
            .OrderByDescending(b => b.TotalRevenue)
            .Take(5)
            .ToList();

        // Expenses by category
        var expensesByCategory = expenses
            .GroupBy(e => e.Kategorie ?? "Sonstige")
            .Select(g => new ExpenseByCategoryDto(
                Category: g.Key,
                Count: g.Count(),
                TotalAmount: g.Sum(e => e.Betrag)
            ))
            .ToList();

        if (purchases.Count > 0)
        {
            expensesByCategory.Add(new ExpenseByCategoryDto(
                Category: "Ankauf",
                Count: purchases.Count,
                TotalAmount: totalPurchaseAmount
            ));
        }

        expensesByCategory = expensesByCategory
            .OrderByDescending(c => c.TotalAmount)
            .ToList();

        return new StatisticsDto(
            StartDate: startDate,
            EndDate: endDate,
            PurchaseCount: purchases.Count,
            SaleCount: sales.Count,
            ExpenseCount: expenses.Count + purchases.Count,
            TotalPurchaseAmount: totalPurchaseAmount,
            TotalSaleAmount: totalSaleAmount,
            TotalExpenseAmount: totalExpenseAmount,
            Profit: profit,
            NetProfit: netProfit,
            AveragePurchasePrice: avgPurchase,
            AverageSalePrice: avgSale,
            AverageProfit: avgProfit,
            DailyBreakdown: dailyBreakdown,
            TopBrands: topBrands,
            ExpensesByCategory: expensesByCategory
        );
    }
}
